src/mongocxx/test/client_side_encryption.cpp

```diff
@@ -1004,20 +1071,79 @@ void _round_trip(class client_encryption* client_encryption,
     REQUIRE(decrypted_val == to_encrypt);
 }
 
-void _run_endpoint_test(class client_encryption* client_encryption,
+void _run_endpoint_test(mongocxx::client* setup_client,
                         bsoncxx::document::view masterkey,
-                        stdx::optional<std::string> error_msg = stdx::nullopt) {
+                        std::string kms_provider,
+                        stdx::optional<std::string> error_msg = stdx::nullopt,
+                        stdx::optional<std::string> invalid_error_msg = stdx::nullopt) {
     INFO("masterkey" << bsoncxx::to_json(masterkey));
 
+    mongocxx::options::client_encryption ce_opts;
+    mongocxx::options::client_encryption ce_opts_invalid;
+    bsoncxx::builder::basic::document kms_doc;
+    bsoncxx::builder::basic::document kms_doc_invalid;
+
+    kms_doc.append(kvp("aws", [&](sub_document subdoc) {
+        subdoc.append(kvp("secretAccessKey",
+                          test_util::getenv_or_fail("MONGOCXX_TEST_AWS_SECRET_ACCESS_KEY")));
+        subdoc.append(
+            kvp("accessKeyId", test_util::getenv_or_fail("MONGOCXX_TEST_AWS_ACCESS_KEY_ID")));
+    }));
+
+    kms_doc.append(kvp("azure", [&](sub_document subdoc) {
+        subdoc.append(kvp("tenantId", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_TENANT_ID")));
+        subdoc.append(kvp("clientId", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_CLIENT_ID")));
+        subdoc.append(
+            kvp("clientSecret", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_CLIENT_SECRET")));
+        subdoc.append(kvp("identityPlatformEndpoint", "login.microsoftonline.com:443"));
+    }));
+
+    kms_doc.append(kvp("gcp", [&](sub_document subdoc) {
+        subdoc.append(kvp("email", test_util::getenv_or_fail("MONGOCXX_TEST_GCP_EMAIL")));
+        subdoc.append(kvp("privateKey", test_util::getenv_or_fail("MONGOCXX_TEST_GCP_PRIVATEKEY")));
+        subdoc.append(kvp("endpoint", "oauth2.googleapis.com:443"));
+    }));
+
+    ce_opts.key_vault_client(setup_client);
+    ce_opts.key_vault_namespace({"keyvault", "datakeys"});
+    ce_opts.kms_providers(kms_doc.view());
+    mongocxx::client_encryption client_encryption{ce_opts};
+
+    kms_doc_invalid.append(kvp("azure", [&](sub_document subdoc) {
+        subdoc.append(kvp("tenantId", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_TENANT_ID")));
+        subdoc.append(kvp("clientId", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_CLIENT_ID")));
+        subdoc.append(
+            kvp("clientSecret", test_util::getenv_or_fail("MONGOCXX_TEST_AZURE_CLIENT_SECRET")));
+        subdoc.append(kvp("identityPlatformEndpoint", "example.com:443"));
```

Comment:
Done-- in the process of doing this, I noticed that there are some other unit tests still using "example.com". I'm happy to change them as part of this change, or just raise awareness as to their being there (client.cpp, pool.com, uri.cpp).

Reply:
I think the other tests that use `example.com` do not need to be updated.

`"A client connects to a provided mongodb uri"` tests that the URI `"mongodb://mongodb.example.com:9999"` is passed to libmongoc. Despite the name, it does not attempt to connect.

`"a pool is created with the correct MongoDB URI"` similarly tests that the URI `"mongodb://mongodb.example.com:9999"` is passed to libmongoc.

The `"Valid URI"` and `"Invalid URI"` tests test URI parsing.

Since the tests do not connect to example.com, I think it is OK to leave them as-is.